include/ignition/common/detail/Plugin.hh

```diff
@@ -15,98 +15,4 @@
  *
  */
 
```

Comment:
Since the `detail` directories aren't part of the public API, I think we can just remove them. 

If `include/ignition/common/Plugin.hh` isn't including `include/ignition/common/detail/Plugin.hh`, no one else is.

This is valid for all libraries.

Reply:
Can I double check if this means that I delete all detail redirection headers across all libraries?

Reply:
> Since the `detail` directories aren't part of the public API, I think we can just remove them.
> 
> If `include/ignition/common/Plugin.hh` isn't including `include/ignition/common/detail/Plugin.hh`, no one else is.
> 
> This is valid for all libraries.

I disagree. These are installed header files, so they are part of the public API, and they are included from the header files not in the `detail` folder, though often from the bottom of the file so it is easy to miss:

* https://github.com/ignitionrobotics/ign-math/blob/ignition-math6_6.10.0/include/ignition/math/Box.hh#L224
* https://github.com/ignitionrobotics/ign-common/blob/ignition-common4_4.5.0/include/ignition/common/Plugin.hh#L194

I think it would be best to keep the redirection headers for all the contents of the `detail`  subfolders

Reply:
> These are installed header files, so they are part of the public API,

I understand that we've never explicitly declared our public API, and we should. But the reason that the `detail` directories exist is that they're not meant to be directly used by downstream. They're installed because of technicalities of the language which require those templates to be installed, but that's an implementation _detail_, not our API. And as such, we don't need to tick-tock them.

ROS QL1 packages [explicitly call this out](https://github.com/ros2/rclcpp/blob/master/rclcpp/QUALITY_DECLARATION.md#public-api-declaration-1iii). I ticketed this issue so we document this properly for Gazebo: 

* https://github.com/ignitionrobotics/docs/issues/248

> I think it would be best to keep the redirection headers for all the contents of the detail subfolders

Having them there wouldn't break anything, but those should be dead files that no one is using. I'd say that keeping them there sends the wrong message that those are part of our public API and subject to tick-tock.

Reply:
I'm getting a very different impression from your first and most recent comments. I think your points are reasonable, but the way they were initially phrased was a bit alarming to me.

> Since the `detail` directories aren't part of the public API, I think we can just remove them.

I've gathered from your subsequent comment that the context is specifically about redirection headers in a `detail`  subfolder, but on first glance it sounded to me like you were saying we can delete `detail` headers at any time from anywhere with no impact on API, which is absolutely not true. I now think you meant we don't need to tick-tock them, which is reasonable.

> If `include/ignition/common/Plugin.hh` isn't including `include/ignition/common/detail/Plugin.hh`, no one else is.

Again, I now gather that this context is about redirection headers, not public headers in general, since `gz/common/Plugin.hh` does include `gz/common/detail/Plugin.hh`

> This is valid for all libraries.

Nothing specific here but it added to my alarm! 😅

Reply:
Ah sorry I should have been clearer. Yes, what I meant is that we can remove the `ignition/*/detail/*.hh` headers that just include `gz/*/detail/*.hh` ✅

Reply:
I think downstream users are less likely to need the `ignition/*/detail/*` redirection headers. I would lean towards redirecting all headers for now until we've migrated headers for all the ignition libraries and then consider removing the `detail` redirects when we add the deprecation warnings to the primary redirection headers.

Reply:
> I think downstream users are less likely to need the ignition/*/detail/* redirection headers

If they're including them directly, that's a mistake on their part.

> I would lean towards redirecting all headers for now until we've migrated headers

I'm ok with that if it makes things easier. We already have a long list of things to go back to, and long PRs to review. If the script could be updated to skip the `detail` folders automatically I think that would make our lives easier...

Reply:
Done! 7b17e475504d5ad308b943e3e5f848577b65d2de